openownership/data-standard
# 0.4.0 schema redesign, JSON Schema version upgrade, and test refactor

# Overview

- What does this pull request do?
  * Changes the schema to match the change over time, declarations, and metadata proposals (https://github.com/openownership/data-standard/issues/487)
  * Adds validation features now available in JSON Schema 2020-12 (#516)
  * Refactors the tests into schema structure tests (test_schema), schema validation tests (test_data) and example data tests (test_docs). Makes use of pytest fixtures for reusable components (#493).

- How can a reviewer test or examine your changes?
  * Run the tests (`pytest tests`) - currently 28 tests in test_docs fail, because the old example data remains, which is all BODS version 0.3.0, but all other tests should pass.
  * Take data which is known to be valid 0.4.0 data, and covers the substantial semantic/structural changes that have been made to the schema, and put it in the tests/data/valid-statements directory, and run `tests/test_data.py` to confirm it validates as expected.

- Who is best placed to review it?
  * Kadie / Kathryn for the schema changes.
  * Ed for the tests refactor.

## TODOs

The following are not completed by this PR, but I don't think they need to block it at this stage.

* There are some todos remaining in the tests relating to bugs (all have open issues to track).
* The _invalid_ data tests are not complete - it's not equipped to check the correct error messages are being generated. 
* There is not enough test data to check all branches of the schema in data/valid-statements yet - we can keep adding this iteratively in smaller batches once this PR is merged.
* The old example data needs to be replaced - we can also do this in smaller batches.
* There are some further changes we can make for #516 which are documented in the spreadsheet, but which we can do as and when in smaller PRs.
* There's a test stubbed to check if the documentation builds - I'll add this at another point in future.
* The documentation fails to build - I'll fix this next, it might be a big job.

## Documentation & Release

- Discussed with Kadie about the changelog update.
- Documentation update is happening in other branches which will all eventually be merged into main with this.
